<a href="https://colab.research.google.com/github/ralucapestritu/ML_Seminar_Pestritu-BoborocRaluca/blob/main/Tema_Laborator_1_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/cristinaboboc-1975/Seminar-Machine-learning-in-Python/blob/main/Tema_Laborator_1_ML.ipynb)

# 📝 Tema de Casă / Exercițiu Individual — Laborator 1
**Curs:** Machine Learning în Python  
**Subiect:** Manipularea datelor cu Python, NumPy și Pandas  
**Timp estimat:** 30–45 minute | **Punctaj total:** 10 puncte

---

### 🎯 Contextul Problemei
Sunteți Data Scientist într-un proiect educațional. Ați primit un set de date care conține rezultatele mai multor studenți la 3 evaluări intermediare (Test 1, Test 2, Proiect), precum și numărul de ore studiate săptămânal.

Obiectivul vostru este să analizați și să pregătiți aceste date pentru un viitor model de Machine Learning.

### 🔹 Cerința 1: Funcții Python & Logică (2 puncte)
Scrieți o funcție Python numită `calculeaza_nota_finala(t1, t2, proiect)` care primește 3 note (numere reale între 1 și 10) și calculează media ponderată după următoarea regulă:
- **Test 1:** pondere de 30% (`0.30`)
- **Test 2:** pondere de 30% (`0.30`)
- **Proiect:** pondere de 40% (`0.40`)

*Cerință:* Funcția trebuie să returneze media ponderată rotunjită la două zecimale. Testați funcția pentru notele: `Test 1 = 8.5`, `Test 2 = 7.0`, `Proiect = 9.5`.

In [9]:
# Rezolvare Cerința 1
def calculeaza_nota_finala(t1, t2, proiect):
    # Scrieți codul aici
    nota_finala=t1*0.3+t2*0.3+proiect*0.4
    return round(nota_finala,2)

# Testare funcție
print(calculeaza_nota_finala(8.5, 7.0, 9.5))

8.45


### 🔹 Cerința 2: Operare cu Vectori în NumPy (3 puncte)
Avem următorul vector NumPy care reprezintă orele de studiu săptămânale ale 8 studenți:
```python
import numpy as np
ore_studiu = np.array([12, 5, 18, 8, 15, 3, 22, 10])
```

1. Calculați **media**, **deviația standard**, valoarea **minimă** și valoarea **maximă** a orelor de studiu folosind funcții din `numpy`.
2. Creați un filtru boolean care să identifice studenții care au studiat **mai mult de 10 ore** pe săptămână.
3. Afișați doar orele de studiu ale studenților care îndeplinesc această condiție.

In [10]:
import numpy as np

ore_studiu = np.array([12, 5, 18, 8, 15, 3, 22, 10])

# 1. Statistici descriptive
media=np.mean(ore_studiu)
deviatia_std=np.std(ore_studiu)
minim=np.min(ore_studiu)
maxim=np.max(ore_studiu)

print(f"media:{media}")
print(f"deviatia standard:{deviatia_std:2f}")
print(f"valoarea minima:{minim}")
print(f"valoarea maxima:{maxim}")

# 2 & 3. Filtrare booleană
filtru_peste_10=ore_studiu>10

ore_filtrate=ore_studiu[filtru_peste_10]
print("ore de studiu mai mari ca 10 ore:", ore_filtrate)

media:11.625
deviatia standard:6.060889
valoarea minima:3
valoarea maxima:22
ore de studiu mai mari ca 10 ore: [12 18 15 22]


### 🔹 Cerința 3: Creare și Manipulare Tabel Pandas (4 puncte)
Folosind dicționarul oferit, creați un `DataFrame` Pandas numit `df_studenti` și realizați următoarele prelucrări:

1. Adăugați o coloană nouă în tabel, numită `Nota_Finala`, calculată aplicând funcția `calculeaza_nota_finala` creată la **Cerința 1** peste coloanele `Test_1`, `Test_2` și `Proiect`.
2. Adăugați o coloană booleană `Admis` care conține valoarea `True` dacă `Nota_Finala >= 6.0` și `False` în caz contrar.
3. Filtrați tabelul și afișați doar studenții **Admiși** care au învățat **peste 10 ore** (`Ore_Studiu > 10`).

In [11]:
# Rezolvare Cerința 3
import pandas as pd

date_studenti = {
    'Nume': ['Ana', 'Bogdan', 'Cristian', 'Elena', 'Florin', 'Gabriela', 'Horia', 'Ioana'],
    'Test_1': [8.5, 5.0, 9.0, 6.5, 4.0, 9.5, 7.0, 8.0],
    'Test_2': [7.0, 6.0, 8.5, 5.5, 4.5, 10.0, 6.5, 8.5],
    'Proiect': [9.5, 7.0, 9.5, 8.0, 5.0, 9.5, 7.5, 9.0],
    'Ore_Studiu': [12, 5, 18, 8, 15, 3, 22, 10]
}

#calcul nota_finala

df_studenti = pd.DataFrame(date_studenti)
df_studenti['nota_finala']=df_studenti.apply(lambda row:calculeaza_nota_finala(row['Test_1'],row['Test_2'],row['Proiect']),axis=1)


# 2. Adăugare coloană Admis
df_studenti['admis']=df_studenti['nota_finala']>=6.0


# 3. Filtrare

studenti_filtrati=df_studenti[(df_studenti['admis']==True)& (df_studenti['Ore_Studiu']>10)]

print("tabel complet")
print(df_studenti)

print("studenti admisi cu peste 10 ore de studiu")
print(studenti_filtrati)


tabel complet
       Nume  Test_1  Test_2  Proiect  Ore_Studiu  nota_finala  admis
0       Ana     8.5     7.0      9.5          12         8.45   True
1    Bogdan     5.0     6.0      7.0           5         6.10   True
2  Cristian     9.0     8.5      9.5          18         9.05   True
3     Elena     6.5     5.5      8.0           8         6.80   True
4    Florin     4.0     4.5      5.0          15         4.55  False
5  Gabriela     9.5    10.0      9.5           3         9.65   True
6     Horia     7.0     6.5      7.5          22         7.05   True
7     Ioana     8.0     8.5      9.0          10         8.55   True
studenti admisi cu peste 10 ore de studiu
       Nume  Test_1  Test_2  Proiect  Ore_Studiu  nota_finala  admis
0       Ana     8.5     7.0      9.5          12         8.45   True
2  Cristian     9.0     8.5      9.5          18         9.05   True
6     Horia     7.0     6.5      7.5          22         7.05   True


### 🔹 Cerința 4: Întrebare de Reflexie (1 punct)
Privind tabelul final obținut:  
*Existe vreun student care a obținut o notă mare (ex. peste 9.0), dar a învățat un număr foarte mic de ore (ex. sub 5 ore)? Cum ați explica acest caz dacă ar fi vorba despre un set de date din lumea reală și ce impact ar avea un astfel de caz asupra unui model de Machine Learning?*

Gabriela a invatat doar 3 ore pe saptamana si a obtinut o nota finala de 9.65.
Exista multiple cauza care pot duce la aceasta situatie.
In primul rand, exista posibilitatea ca nota sa nu fie intr-adevar reala si Gabriela sa fi copiat.
Daca Gabriela a fost un student corect, exista alte posibilitati valabile. Spre exemplu, Gabriela avea deja cunostinde de machine learning sau are o capacitatea de invatare si memorare de exceptie.

Impactul asupra unui model de machine learning acest caz reprezinta un outlier. El poate deruta modelul facandu-l sa considere ca orele de studiu nu afecteaza performanta. In acest caz sunt doua optiuni: eliminarea sau pastrarea si adaptarea modelului.

Eliminarea lui imbunatateste performanta in cazul unui model simplu, precum cel de regresie si elimina zgomotul din date. Dar, eliminarea duce la reducerea brusca a dimensiunii datelor deoarece 1 din 8 studenti reprezinta foarte mult, setul de date fiind foarte mic. In plus, apare riscul de subreprezentativitate, acest tip de studenti fiind totusi cazuri care pot aparea in realitate.

In cazul in care valoarea se pastreaza si se foloseste un model mai robust la outliers, precum Random Forest, se va capta mai bine complexitatea reala, iar modelul va fi antrenat pe situatii speciale. Totusi, pastrarea poate duce la overfitting, iar o tratare corecta ar trebui sa includa atribute noi (nivelul de cunostinte anterioare), situatie care necesita timp de culegere si prelucrare mai indelungat.